pkg/snet/stun_handler.go

```diff
@@ -0,0 +1,386 @@
+	}
+	handler.cond = sync.NewCond(&handler.mutex)
+	return handler, nil
+}
+
+func (c *stunHandler) queuePacket(pkt bufferedPacket) bool {
+	if c.queuedBytes.Load()+int64(len(pkt.data)) > c.maxQueuedBytes {
+		return false
+	}
+	select {
+	case c.recvChan <- pkt:
+		c.queuedBytes.Add(int64(len(pkt.data)))
+		return true
+	default:
+		return false
+	}
+}
+
+func (c *stunHandler) dequeuePacket() (bufferedPacket, bool) {
+	select {
+	case pkt := <-c.recvChan:
+		c.queuedBytes.Add(-int64(len(pkt.data)))
+		return pkt, true
+	default:
+		return bufferedPacket{}, false
+	}
+}
+
+func (c *stunHandler) ReadFrom(b []byte) (int, net.Addr, error) {
+	for {
+		pkt, ok := c.dequeuePacket()
+		if ok {
+			copy(b, pkt.data)
+			return pkt.len, pkt.addr, nil
+		}
+		n, addr, err := c.UDPConn.ReadFrom(b)
+		if err != nil {
+			return n, addr, err
+		}
+		if stun.Is(b) {
+			c.recvStunChan <- b[:n]
+		} else {
+			return n, addr, nil
+		}
+	}
+}
+
+func (c *stunHandler) readStunPacket(ctx context.Context) ([]byte, error) {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
+			select {
+			case pkt := <-c.recvStunChan:
+				return pkt, nil
+			default:
+				buf := make([]byte, 1500)
+				n, addr, err := c.UDPConn.ReadFrom(buf)
+				if err != nil {
+					return nil, err
+				}
+				if stun.Is(buf[:n]) {
+					return buf[:n], nil
+				} else {
+					c.queuePacket(bufferedPacket{data: buf[:n], addr: addr, len: n})
+				}
+			}
+		}
+	}
+}
+
+func (c *stunHandler) getStunResponse(ctx context.Context, txid stun.TxID) (netip.AddrPort, error) {
+	c.mutex.Lock()
+	ch, ok := c.stunChans[txid]
+	c.mutex.Unlock()
+	if !ok {
+		return netip.AddrPort{}, fmt.Errorf("STUN response channel not found")
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			return netip.AddrPort{}, ctx.Err()
+		default:
+			select {
+			case resp := <-ch:
+				return resp.mappedAddr, resp.err
+			default:
+				pkt, err := c.readStunPacket(ctx)
+				if err != nil {
+					return netip.AddrPort{}, err
+				}
+				respTxID, mappedAddr, err := stun.ParseResponse(pkt)
+				if err != nil {
+					continue
+				}
+				if respTxID == txid {
+					return mappedAddr, nil
+				}
+				// Send to the appropriate channel
+				c.mutex.Lock()
+				if ch, ok := c.stunChans[respTxID]; ok {
+					ch <- stunResponse{mappedAddr: mappedAddr, err: err}
+				}
+				c.mutex.Unlock()
+			}
+		}
+	}
+}
+
+type natMapping struct {
+	destination *net.UDPAddr
+	mappedAddr  *net.UDPAddr
+	lastUsed    time.Time
+}
+
+func (mapping *natMapping) touch() {
+	mapping.lastUsed = time.Now()
+}
+
+func (mapping *natMapping) isValid() bool {
+	return time.Since(mapping.lastUsed) < timeoutDuration
+}
+
+func (c *stunHandler) getMappedAddr(dest *net.UDPAddr) (*net.UDPAddr, error) {
+	c.mutex.Lock()
+	for {
+		// Check if mapping exists and is valid
+		if mapping, ok := c.mappings[dest]; ok {
+			if mapping.isValid() {
+				mapping.touch()
+				c.mutex.Unlock()
+				return mapping.mappedAddr, nil
```

Comment:
This may cause some racing problem when accessing a field of the struct after releasing the lock, adding:
```
if mapping.isValid() {
      mapping.touch()
      addr := mapping.mappedAddr 
      c.mutex.Unlock() 
      return addr, nil
 }
```
should avoid it.